src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataPrimitiveSerializer.cs

```diff
@@ -155,19 +156,54 @@ internal static object ConvertPrimitiveValue(object value, IEdmPrimitiveTypeRefe
             }
 
             Type type = value.GetType();
-            if (primitiveType != null && primitiveType.IsDate() && TypeHelper.IsDateTime(type))
+
+            if (primitiveType != null)
             {
-                Date dt = (DateTime)value;
-                return dt;
+                if (primitiveType.IsDate() && TypeHelper.IsDateTime(type))
+                {
+                    Date dt = (DateTime)value;
+                    return dt;
+                }
+
+                if (primitiveType.IsTimeOfDay() && TypeHelper.IsTimeSpan(type))
+                {
+                    TimeOfDay tod = (TimeSpan)value;
+                    return tod;
+                }
+
+                if (primitiveType is EdmDecimalTypeReference decimalTypeReference && decimalTypeReference.Scale.HasValue && value is decimal decimalValue)
+                {
+                    return CalculateScaleOfDecimal(decimalValue, decimalTypeReference.Scale.Value);
+                }
             }
 
-            if (primitiveType != null && primitiveType.IsTimeOfDay() && TypeHelper.IsTimeSpan(type))
+            return ConvertUnsupportedPrimitives(value);
+        }
+
+        static decimal CalculateScaleOfDecimal(decimal value, int scale)
+        {
+            string leftSide = GetLeftSideFromPoint(value);
+            string rightSide = GetRightSideFromPoint(value);
+            var numberOfZeroToAddToTheEndOfDecimalValue = scale - rightSide.Length;
+            for (int i = 0; i < numberOfZeroToAddToTheEndOfDecimalValue; i++)
             {
-                TimeOfDay tod = (TimeSpan)value;
-                return tod;
+                rightSide += "0";
             }
+            value = decimal.Parse($"{leftSide}.{rightSide}", CultureInfo.InvariantCulture);
```

Comment:
Why is this necessary? Why is calling `decimal.Round(...)` not sufficient? Is this to ensure that we always have `scale` number of digits after the decimal point by padding `0`'s at the end if the original value has fewer digits?

If that's the case, I feel like that's unnecessary work. According to the [spec](https://docs.oasis-open.org/odata/odata-csdl-json/v4.01/odata-csdl-json-v4.01.html#sec_Scale), Scale specifies the **maximum** number of digits allowed after the decimal point, **not** the **exact** number of digits. So a value like `10.1` is valid for Scale = 2, we don't need to coerce it to `10.10` in order to be valid.

If the service requires that values have an exact number of decimal places, maybe the service should ensure the values adhere to that specification before being processed by OData. By forcing an exact number of decimal places, I think we would be deviating from the expectations of the library. I think we should do `decimal.Round(value, scale, ...)` and leave the rest up to the service or client. My 2 cents.